internal/client/ssh.go

```diff
@@ -0,0 +1,386 @@
+}
+
+func (c *SSHClient) determineOS() (string, error) {
+	session, err := c.newSSHSession()
+	if err != nil {
+		return "", err
+	}
+	defer session.Close()
+
+	output, err := session.Output("cat /etc/os-release | grep -w ID=")
+	if err != nil {
+		return "", fmt.Errorf("failed to determine os: %s", string(output))
+	}
+
+	os := strings.ReplaceAll(string(output), "ID=", "")
+	os = strings.ReplaceAll(os, "\"", "")
+	return strings.TrimSpace(os), nil
+}
+
+func (c *SSHClient) Connect() error {
+	var err error
+	var authMethods []ssh.AuthMethod
+	var hostkeyCallback ssh.HostKeyCallback
+
+	if c.PKPath != "" {
+		pkPath := c.PKPath
+		// resolve home directory
+		if strings.HasPrefix(pkPath, "~/") {
+			homeDir, err := os.UserHomeDir()
+			if err != nil {
+				return err
+			}
+			pkPath = strings.Replace(pkPath, "~/", homeDir+"/", 1)
+		}
+		pKey, err := os.ReadFile(pkPath)
+		if err != nil {
+			return err
+		}
+
+		signer, err := ssh.ParsePrivateKey(pKey)
+		if err != nil {
+			return err
+		}
+
+		authMethods = append(authMethods, ssh.PublicKeys(signer))
+	}
+
+	if c.Password != "" {
+		authMethods = append(authMethods, ssh.Password(c.Password))
+	}
+
+	hostkeyCallback = ignoreHostKeyCallback
+	if !c.InsecureIgnoreHostKey {
+		homeDir, err := os.UserHomeDir()
+		if err != nil {
+			return err
+		}
+
+		knownHostsPath := fmt.Sprintf("%s/.ssh/known_hosts", homeDir)
+		hostkeyCallback, err = knownhosts.New(knownHostsPath)
+		if err != nil {
+			return err
+		}
+	}
+
+	conf := &ssh.ClientConfig{
+		User:            c.User,
+		HostKeyCallback: hostkeyCallback,
+		Auth:            authMethods,
+	}
+
+	c.conn, err = ssh.Dial("tcp", c.Host, conf)
+	if err != nil {
+		return err
+	}
+
+	_, err = c.hasRootPrivileges()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *SSHClient) Disconnect() error {
+	if c == nil {
+		return nil
+	}
+
+	if c.conn != nil {
+		if err := c.conn.Close(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *SSHClient) InstallBlueChi(installCtrl bool, installAgent bool) error {
+	needsInstallCtrl := false
+	needsInstallAgent := false
+
+	if installCtrl {
+		isInstalled, err := c.isServiceInstalled("bluechi-controller.service")
+		if err != nil {
+			return err
+		}
+		needsInstallCtrl = !isInstalled
+	}
+	if installAgent {
+		isInstalled, err := c.isServiceInstalled("bluechi-agent.service")
+		if err != nil {
+			return err
+		}
+		needsInstallAgent = !isInstalled
+	}
+
+	if !needsInstallCtrl && !needsInstallAgent {
+		return nil
+	}
+
+	os, err := c.determineOS()
+	if err != nil {
+		return err
+	}
+
+	if os == "autosd" || os == "centos" {
+		packagesToInstall := ""
+		if needsInstallCtrl {
+			packagesToInstall += " bluechi-controller bluechi-ctl "
+		}
+		if needsInstallAgent {
+			packagesToInstall += " bluechi-agent"
+		}
+
+		session, err := c.newSSHSession()
+		if err != nil {
+			return err
+		}
+		defer session.Close()
+
+		sudoPrefix := ""
+		if c.connHasRoot {
```

Comment:
Maybe I'm missing something here. From what I can see, `c.connHasRoot` is `true` if the `whomai` returned `root`. So, shouldn't this check be the opposite?
The same question applies to other uses of `c.connHasRoot`

Reply:
Yes, this is a bug. Fixed it.